File: typescript/core/src/ChunkBuilder.ts
import { McapRecordBuilder } from "./McapRecordBuilder.ts";
import type { Channel, Message, MessageIndex, Schema } from "./types.ts";

type ChunkBuilderOptions = {
  useMessageIndex?: boolean;
};
class ChunkBuilder {
  #recordWriter = new McapRecordBuilder();
  #messageIndices: Map<number, MessageIndex> | undefined;
  #totalMessageCount = 0;

  messageStartTime = 0n;
  messageEndTime = 0n;

  constructor({ useMessageIndex = true }: ChunkBuilderOptions) {
    if (useMessageIndex) {
      this.#messageIndices = new Map();
    }
  }

  get numMessages(): number {
    return this.#totalMessageCount;
  }

  get buffer(): Uint8Array {
    return this.#recordWriter.buffer;
  }

  get byteLength(): number {
    return this.#recordWriter.length;
  }

  get indices(): Iterable<MessageIndex> {
    if (this.#messageIndices) {
      return this.#messageIndices.values();
    }
    return [];
  }

  addSchema(schema: Schema): void {
    this.#recordWriter.writeSchema(schema);
  }

  addChannel(info: Channel): void {
    if (this.#messageIndices && !this.#messageIndices.has(info.id)) {
      this.#messageIndices.set(info.id, {
        channelId: info.id,
        records: [],
      });
    }
    this.#recordWriter.writeChannel(info);
  }

  addMessage(message: Message): void {
    if (this.#totalMessageCount === 0 || message.logTime < this.messageStartTime) {
      this.messageStartTime = message.logTime;
    }
    if (this.#totalMessageCount === 0 || message.logTime > this.messageEndTime) {
      this.messageEndTime = message.logTime;
    }

    if (this.#messageIndices) {
      let messageIndex = this.#messageIndices.get(message.channelId);
      if (!messageIndex) {
        messageIndex = {
          channelId: message.channelId,
          records: [],
        };
        this.#messageIndices.set(message.channelId, messageIndex);
      }
      messageIndex.records.push([message.logTime, BigInt(this.#recordWriter.length)]);
    }

    this.#totalMessageCount += 1;
    this.#recordWriter.writeMessage(message);
  }

  reset(): void {
    this.messageStartTime = 0n;
    this.messageEndTime = 0n;
    this.#totalMessageCount = 0;
    this.#messageIndices?.clear();
    this.#recordWriter.reset();
  }
}

export { ChunkBuilder };

File: python/mcap/mcap/_chunk_builder.py
from typing import Dict

from .data_stream import RecordBuilder
from .records import Channel, Message, MessageIndex, Schema


class ChunkBuilder:
    def __init__(self) -> None:
        self.message_end_time = 0
        self.message_indices: Dict[int, MessageIndex] = {}
        self.message_start_time = 0
        self.record_writer = RecordBuilder()
        self.num_messages = 0

    @property
    def count(self):
        return self.record_writer.count

    def end(self):
        return self.record_writer.end()

    def add_channel(self, channel: Channel):
        channel.write(self.record_writer)

    def add_schema(self, schema: Schema):
        schema.write(self.record_writer)

    def add_message(self, message: Message):
        if self.num_messages == 0:
            self.message_start_time = message.log_time
        else:
            self.message_start_time = min(self.message_start_time, message.log_time)
        self.message_end_time = max(self.message_end_time, message.log_time)

        if not self.message_indices.get(message.channel_id):
            self.message_indices[message.channel_id] = MessageIndex(
                channel_id=message.channel_id, records=[]
            )
        self.message_indices[message.channel_id].records.append(
            (message.log_time, self.record_writer.count)
        )

        self.num_messages += 1
        message.write(self.record_writer)

    def reset(self):
        self.message_end_time = 0
        self.message_indices.clear()
        self.message_start_time = 0
        self.record_writer.end()
        self.num_messages = 0
